File: scripts/javascript/update_verifier.js
/*
 * update verifier
 */
const { BN, Long, bytes, units } = require('@zilliqa-js/util');
const { Zilliqa } = require('@zilliqa-js/zilliqa');
const { toBech32Address, getAddressFromPrivateKey } = require('@zilliqa-js/crypto');

// change the following parameters
const API = 'http://localhost:5555'
const CHAIN_ID = 1;
const PRIVATE_KEY = 'e53d1c3edaffc7a7bab5418eb836cf75819a82872b4a1a0f1c7fcf5c3e020b89'; // admin
const STAKING_PROXY_ADDR = toBech32Address("0xDB5Dc7118765A84B6c6A582280fA37A1DD2d9f69"); // checksum proxy address
const NEW_VERIFIER = '0x381f4008505e940ad7681ec3468a719060caf796'; // new verifier checksum address with '0x'

const zilliqa = new Zilliqa(API);
const MSG_VERSION = 1;
const VERSION = bytes.pack(CHAIN_ID, MSG_VERSION);
const GAS_PRICE = units.toQa('1000', units.Units.Li);


async function main() {
    zilliqa.wallet.addByPrivateKey(PRIVATE_KEY);
    const address = getAddressFromPrivateKey(PRIVATE_KEY);
    console.log("Your account address is: %o", `${address}`);
    console.log("proxy: %o\n", STAKING_PROXY_ADDR);

    console.log("------------------------ begin update verifier ------------------------\n");
    try {
        const contract = zilliqa.contracts.at(STAKING_PROXY_ADDR);
        const callTx = await contract.call(
            'update_verifier',
            [
                {
                    vname: 'verif',
                    type: 'ByStr20',
                    value: `${NEW_VERIFIER}`
                }
            ],
            {
                version: VERSION,
                amount: new BN(0),
                gasPrice: GAS_PRICE,
                gasLimit: Long.fromNumber(10000)
            },
            33,
            1000,
            true
        );
        console.log("transaction: %o", callTx.id);
        console.log(JSON.stringify(callTx.receipt, null, 4));

    } catch (err) {
        console.log(err);
    }
    console.log("------------------------ end update verifier ------------------------\n");
}

main();

File: phase0/tests/transitions/update_verifier.go
package transitions

import (
	"encoding/json"
	"errors"
	"fmt"
	"github.com/Zilliqa/gozilliqa-sdk/bech32"
	contract2 "github.com/Zilliqa/gozilliqa-sdk/contract"
	"github.com/Zilliqa/gozilliqa-sdk/keytools"
	"github.com/Zilliqa/gozilliqa-sdk/util"
	"strings"
)

func (p *Proxy) UpdateVerifier(valid, invalid string) {
	err := p.updateVerifier(valid)
	if err != nil {
		panic("update verifier with admin failed: " + err.Error())
	}
	fmt.Println("update verifier with admin succeed")

	err2 := p.updateVerifier(invalid)
	if err2 == nil {
		panic("update verifier with invalid admin failed")
	}
	fmt.Println("update verifier with invalid admin succeed")
}

func (p *Proxy) updateVerifierTo(private string, newVerifier string) error {
	proxy, _ := bech32.ToBech32Address(p.Addr)

	parameters := []contract2.Value{
		{
			VName: "verif",
			Type:  "ByStr20",
			Value: newVerifier,
		},
	}
	args, _ := json.Marshal(parameters)
	if err2, output := ExecZli("contract", "call",
		"-k", private,
		"-a", proxy,
		"-t", "update_verifier",
		"-f", "true",
		"-r", string(args)); err2 != nil {
		return errors.New("call transition error: " + err2.Error())
	} else {
		tx := strings.TrimSpace(strings.Split(output, "confirmed!")[1])
		fmt.Println("transaction id = ", tx)
		payload := p.Provider.GetTransaction(tx).Result.(map[string]interface{})
		receipt := payload["receipt"].(map[string]interface{})
		success := receipt["success"].(bool)
		if success {
			res := p.Provider.GetSmartContractState(p.ImplAddress).Result.(map[string]interface{})
			verifier := res["verifier"].(map[string]interface{})
			arguments := verifier["arguments"].([]interface{})[0]
			fmt.Println(arguments)
			if arguments == nil {
				return errors.New("verifier is none")
			}
			arg := arguments.(string)
			if arg == newVerifier {
				return nil
			} else {
				return errors.New("update state failed")
			}
		} else {
			return errors.New("transaction failed")
		}
	}
}

func (p *Proxy) updateVerifier(private string) error {
	operator := "0x" + keytools.GetAddressFromPrivateKey(util.DecodeHex(private))
	return p.updateVerifierTo(private, operator)
}
